book/optimization.qmd

```diff
@@ -772,12 +592,404 @@ It is therefore important to ensure that the performance of a tuned model is *al
+
+```{r optimization-035}
+search_space = ps(
+  cost   = p_dbl(lower = 0.1, upper = 1),
+  kernel = p_fct(levels = c("polynomial", "radial"))
+)
+search_space
+```
+
+Instead of using the `logscale` option, we can also manually set a transformation (`trafo`).
+
+```{r optimization-037}
+search_space = ps(
+  cost   = p_dbl(-1, 1, trafo = function(x) 10^x),
+  kernel = p_fct(c("polynomial", "radial"))
+)
+```
+
+It is even possible to attach another transformation to the `r ref("ParamSet")` as a whole that gets executed after individual parameter's transformations were performed.
+It is given through the `.extra_trafo` argument and should be a function with parameters `x` and `param_set` that takes a list of parameter values in `x` and returns a modified list.
+This transformation can access all parameter values of an evaluation and modify them with interactions.
+It is even possible to add or remove parameters.
+
+In our example we now assume that the parameter `cost` should be set to a higher value when the `kernel` is `"polynomial."`.
+
+```{r optimization-039}
+search_space = ps(
+  cost = p_dbl(-1, 1, trafo = function(x) 10^x),
+  kernel = p_fct(c("polynomial", "radial")),
+  .extra_trafo = function(x, param_set) {
+    if (x$kernel == "polynomial") {
+      x$cost = x$cost + 2
+    }
+    x
+  }
+)
+```
+
+```{r optimization-040}
+#| echo: false
+#| label: fig-technical-paradox-trafo
+#| fig-cap: Design grid for tuning a SVM. The resolution is 5, the cost parameter is logaithmically transformed when points with a `kernel` equal to `"polynomial"` are shifted to the right by a value of 2.
+dat3 = rbindlist(generate_design_grid(search_space, 5)$transpose())
+theme_set(theme_minimal())
+ggplot(data = dat3, aes(cost, kernel)) +
+    geom_point()
+```
+
+The available types of search space parameters are limited: continuous, integer, discrete, and logical scalars.
+There are many machine learning algorithms, however, that take parameters of other types, for example vectors or functions.
+These can not be defined in a search space `r ref("ParamSet")`, and they are often given as `r ref("ParamUty")` in the `r ref("Learner")`'s `r ref("ParamSet")`.^[Recall that a parameter of type `r ref("ParamUty")` cannot be used in a search space because it is never bounded.]
+When trying to tune over these hyperparameters, it is necessary to perform a Transformation that changes the type of the parameter.
+
+An example is the `class.weights` parameter of the [Support Vector Machine](https://machinelearningmastery.com/cost-sensitive-svm-for-imbalanced-classification/) (SVM), which takes a named vector of class weights with one entry for each target class.
+The transformation that would tune `class.weights` for the `r ref("mlr_tasks_spam", text = "spam")` dataset could be:
+
+```{r optimization-041}
+search_space = ps(
+  class.weights = p_dbl(lower = 0.1, upper = 0.9,
+    trafo = function(x) c(spam = x, nonspam = 1 - x))
+)
+```
```

Comment:
I think this is a bit confusing following from the previous example (also subsequent examples below). Why sometimes `.extra_trafo` and others `trafo`. Why not use vector example again? (these are rhetorical Qs I know the answer) I think the flow could be a bit clearer, maybe something like "We will now look at common advanced tuning methods" then you can talk about vectors, characters, etc. But just framed as problem/solution

Reply:
@sebffischer

Reply:
Maybe a graphic would be nice, which illustrates how the trafos are first applied and then the .extra_trafo is applied to all the transformed parameters?

Reply:
@be-marc

Reply:
I added an explanation